### DISCUSSION TITLE
Simplify comprehensions for optional fields

### DISCUSSION THREAD

**Discussion by istvan-andrasi** [cue-lang/cue#2283] [Proposal]
I'm playing with wrapping docker commands (`build`, `volume`/`secret`/`config` `create`/`remove`, `stack deploy`, ...) and I'm quite happy with what I've got so far.

For context, my simplified definition is:
~~~
#stacks: [stackname=string]: {
  volumes?: [name=string]: define?: {external?: bool, ...}
  configs?: [name=string]: define?: {external?: bool, ...}
  secrets?: [name=string]: define?: {external?: bool, ...}
}
~~~

The next thing that I'd be happy with is a simplified version of this comprehension: (x2 for `configs` and `secrets`)
~~~
if stack.volumes != _|_ {
 for name, volume in stack.volumes if volume.define != _|_ {
   if volume.define.external != _|_ {
     if volume.define.external {
       command: "\(stackname)_create_volume_\(name)"`": {...}
       command: "\(stackname)_remove_volume_\(name)"`": {...}
}}}}
~~~

**I propose** that inside comprehensions a trailing `?` after a field name in field references can replace the explicit testing for bottom, which would result in:
~~~
for name, volume in stack.volumes? if volume.define?.external? {
  command: "\(stackname)_create_volume_\(name)"`": {...}
  command: "\(stackname)_remove_volume_\(name)"`": {...}
}
~~~

I'd like to scope this to comprehensions only as a first step.

**Comment by mpvl** [comment]
We anticipate this syntax more generally as part of the query proposal.

**Reply by istvan-andrasi** [reply]
OK, looking forward to it!

**Reply by mpvl** [reply]
We do have a try proposal out now that I think is relevant for this.

see here: https://github.com/cue-lang/cue/discussions/4019

**Reply by istvan-andrasi** [reply]
Thanks for the heads up! Looks promising! Keep up the good work ;)
